### ISSUE TITLE
How do you create an offline app?

### ISSUE THREAD

**Issue by muppsy007** [calcinai/xero-php#695]
Been following this Issue https://github.com/calcinai/xero-php/issues/625 and the many disgruntled dev threads on Xero's own forums. What I'm trying to do is replace an offline app that simply creates an invoice (and a contact if it doesn't exist). 

The demo.php file in the library works for me, but only once. I also understand that I add offline_access to the scopes to get the refresh token returned. What I'm not seeing here or in Xero is a process flow for how this all fits together. 

1. Do you store the token in a DB or something?
2. If so, how do we use it once it's stored? I can't see any code examples for avoiding stepping through Xero's access grant screens.
3. How long do these refresh tokens last? 30 mins? So do we need a separate task that runs every < 30 mins that maintains the refresh token?
4. How is the refresh token maintained?

**Comment by jamieisps** [comment]
I have managed to achieve this by saving the token, refresh token, tenant and expire time. As you mention I have the exact same simple app to create invoices and contacts for each order. I am storing this in the .env file and updating this is the token is refreshed.

So as you mentioned I do an initial manual process to get the data and store this. Then a new endpoint to either use the existing offline token, or if expired this generated a new token and update the .env.

When using the refresh token you no longer need to manually grant access so it just regenerates automatically.

**Comment by muppsy007** [comment]
Thanks @jamieisps. 

So am I understanding the flow correctly here:

1. Do a manual connection in a browser as an initial one-off. Store the response containing the token, refresh token etc in a database field (sounds like you use a file instead).
2. When a website transaction is triggered, use the stored refresh token to connect and update the token, refresh token etc in your local storage.
3. In the same transaction, use the refresh token to run the invoice/contact creation.

Is that right? If so, my only other remaining question is the lifespan of the refresh token. Docs mention 30 mins lifetime, but also 60 days lifetime. If the website is not dealing with transactions within 30mins, is a separate process required to keep getting the refresh token updated every 29mins (for instance)?

**Comment by rodjsta** [comment]
@muppsy007, your workflow steps are basically correct.

1. Do your initial connection and store the access token, refresh token, access token expiry, tenantId and connectionId
2. Whenever a user/your app wants to make a call to Xero, check that the date/time is within 30 minutes of the access token expiry. If it's not, you will need to use your refresh token to get a new access token
3. Update your storage for all of the tokens/expiry (including refresh)
4. Continue your call to Xero using the most up-to-date access token and tenantId

if ( strtotime($accessTokenExpiry) <= time() ) {

	$provider = new Xero([
		'clientId' => $_ENV['XERO_CLIENT_ID'],
		'clientSecret' => $_ENV['XERO_CLIENT_SECRET'],
    	'redirectUri' => $_ENV['XERO_REDIRECT_URL'],
	]);

	$newAccessToken = $provider->getAccessToken('refresh_token', [
		'grant_type' => 'refresh_token',
		'refresh_token' => $refreshToken
	]);

    $accessToken = $newAccessToken->getToken();
    $accessTokenExpiry = date("Y-m-d H:i:s", $newAccessToken->getExpires());
    $refreshToken = $newAccessToken->getRefreshToken();

    // update DB

}

**Comment by muppsy007** [comment]
Thanks for the assist @rodjsta . Giving this a stab now.

**Comment by outdoorsnz** [comment]
From my understanding your initial token is good for 60 days if you keep refreshing and saving new refreshed token, but just means you need to remember the manual login to load new token before / at 60 days which is a pain if your app has long time use.

> if ( strtotime($accessTokenExpiry) <= time() ) {

[xero-php-auth2](https://github.com/XeroAPI/xero-php-oauth2) website example function getHasExpired() use time() without strtotime which is a bug possibly. Comparing two different times strings.

FYI

I lifted quite a bit of this code (https://github.com/XeroAPI/xero-php-oauth2) and reworked to suit my situation.

**Comment by muppsy007** [comment]
@outdoorsnz 

Are you saying that even with constant refreshing and usage, we still need to go through the manual connection step every two months?

**Comment by outdoorsnz** [comment]
> @outdoorsnz
> 
> Are you saying that even with constant refreshing and usage, we still need to go through the manual connection step every two months?

@muppsy007  muppsy007 
Sorry still working this out also, but looks like a refresh token not used expires after 60 days. So as long as you keep refreshing, you should not need to re authorise.
https://developer.xero.com/faq/oauth2/oauth-refresh-expiry

**Comment by rodrigomanara** [comment]
HI All,

Just one question regards to this issue, I have create a cron job to update the refresh token every 30 minutes, it does work for 3 times then after that it does stop do you know why?

**Comment by Healyhatman** [comment]
You don't need to do that, all you need is the previous refresh token and
the access token. Access tokens run out after 30 minutes but can be
refreshed at any time within 30 days.
